src/polyfills/formdata-entries.js

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {map} from '../utils/object';
+
+/**
+ * Creates a new `FormData` object.
+ *
+ * This patches the native constructor so that the polyfill can capture the
+ * fields in the form passed to the constructor. If `FormData#entries` needs a
+ * polyfill, chances are there are no methods to read the content of the
+ * `FormData` after construction, so the only way to implement `entries`
+ * polyfill is to patch the constructor (and the `append` method).
+ *
+ * For more details on this, see http://mdn.io/FormData.
+ *
+ * @param {!HTMLFormElement=} form An HTML `<form>` element — when specified,
+ *     the FormData object will be populated with the form's current keys/values
+ *     using the name property of each element for the keys and their submitted
+ *     value for the values. It will also encode file input content.
+ */
+function FormDataPolyfill(form) {
+  const instance = new FormDataPolyfill.FormDataNative_(form);
+
+  instance.fieldValues_ = map();
+
+  if (!form) {
+    return instance;
+  }
+
+  for (let i = 0; i < form.elements.length; i++) {
+    const control = form.elements[i];
+
+    if (!control.name || control.disabled) {
+      continue;
+    }
+
+    switch (control.type) {
+      case 'select-multiple':
+      case 'select-one':
+        Array.from(control.selectedOptions).forEach(
```

Comment:
... Turns out that `Array.from` is banned by `gulp check-types`, probably because it's not supported by IE. This means that my ideal unified `Array.from(formData.entries())` API usage won't work on IE.

@choumx Will, do you think it's better to polyfill `Array.from` so that `Array.from(formData.entries())` works on IE, or should we just patch a public `entriesArray` API directly on `FormData.prototype`? If we do the latter, callers would then just use `entriesArray` directly instead of `entries`. Also, on browsers with native `entries`, `entriesArray` would simply return `entries`...

Reply:
Yea, I think that's a better option since it avoids confusion from overriding a native API.

Reply:
We can't polyfill `Array.from`, it's massive and has all kinds of edge cases to handle.

I think we already have an entries helper in `amp-form.js` that we can extract here.

Reply:
Done. Thanks Justin for pointing out the existing code that does the constructor work! Dedup'ed the code.

Also, I just realized that although we can't write generator functions in AMP code, we can [manually implement a generator object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators). Callers would then be able to use the `next()` method on the iterator to manipulate `FormData.prototype.entries`'s return value. As a result, I updated this PR to still polyfill the native `entries`.

Reply:
@choumx @jridgewell Just in case you didn't get notified about the previous comment...